docs/adr/adr-019-forced-inclusion-mechanism.md

```diff
@@ -4,435 +4,697 @@
 
-The forced inclusion mechanism will be configurable with the following parameters:
+### Configuration
 
 ```go
-type ForcedInclusionConfig struct {
-    Enabled                   bool          // Whether forced inclusion is enabled
-    MaxInclusionDelay         uint64        // Maximum time window for transaction inclusion
-    SequencerDownTime         uint64        // Time after which the sequencer is considered down
-    MinDADelay               uint64        // Minimum number of DA blocks before including a direct tx
+type Genesis struct {
+    ChainID                string
+    StartTime              time.Time
+    InitialHeight          uint64
+    ProposerAddress        []byte
+    DAStartHeight          uint64
+    // Number of DA blocks to scan per forced inclusion fetch
+    // Higher values reduce DA queries but increase latency
+    // Lower values increase DA queries but improve responsiveness
+    DAEpochForcedInclusion uint64
+}
+
+type DAConfig struct {
+    // ... existing fields ...
+
+    // Namespace for forced inclusion transactions
+    ForcedInclusionNamespace string
+}
+
+type NodeConfig struct {
+    // ... existing fields ...
+
+    // Run node with based sequencer (requires aggregator mode)
+    BasedSequencer bool
 }
 ```
 
+### Configuration Examples
+
+#### Traditional Sequencer with Forced Inclusion
+
+```yaml
+# genesis.json
+{
+  "chain_id": "my-rollup",
+  "forced_inclusion_da_epoch": 10  # Scan 10 DA blocks at a time
+}
+
+# config.toml
+[da]
+forced_inclusion_namespace = "0x0000000000000000000000000000000000000000000000000000666f72636564"
+
+[node]
+aggregator = true
+based_sequencer = false # Use traditional sequencer
+```
+
+#### Based Sequencer (DA-Only)
+
+```yaml
+# genesis.json
+{
+  "chain_id": "my-rollup",
+  "forced_inclusion_da_epoch": 5  # Scan 5 DA blocks at a time
+}
+
+# config.toml
+[da]
+forced_inclusion_namespace = "0x0000000000000000000000000000000000000000000000000000666f72636564"
+
+[node]
+aggregator = true
+based_sequencer = true # Use based sequencer
+```
+
+### Sequencer Operation Flows
+
+#### Single Sequencer Flow
+
+```
+1. Timer triggers GetNextBatch
+2. Fetch forced inclusion txs from DA (via DA Retriever)
+   - Only at epoch boundaries
+   - Scan epoch range for forced transactions
+3. Get batch from mempool queue
+4. Prepend forced txs to batch
+5. Return batch for block production
+```
+
+#### Based Sequencer Flow
+
+```
+1. Timer triggers GetNextBatch
+2. Check transaction queue for buffered txs
+3. If queue empty or epoch boundary:
+   - Fetch forced inclusion txs from DA
+   - Add to queue
+4. Create batch from queue (respecting MaxBytes)
+5. Return batch for block production
+```
+
+### Full Node Verification Flow
+
+```
+1. Receive block from DA or P2P
+2. Before applying block:
+   a. Fetch forced inclusion txs from DA at block's DA height
+   b. Build map of transactions in block
+   c. Verify all forced txs are in block
```

Comment:
this can be done after the fact right? like if the block with forced inclusion gets created but the full node doesnt have the data then it wont block waiting for the data?